PlayCover/Utils/URLHandler.swift

```diff
@@ -0,0 +1,92 @@
+//
+//  URIHandler.swift
+//  PlayCover
+//
+//  Created by Venti on 14/02/2023.
+//
+
+import Foundation
+
+enum URLTypes: Int, Equatable {
+    case source
+    case keymap
+    case app
+}
+
+enum URLAction: Int, Equatable {
+    case add
+    case remove
+    case update
+    case install
+    case open
+}
+
+class URLObservable: ObservableObject {
+    @Published var url: String?
+    @Published var type: URLTypes?
+    @Published var action: URLAction?
+
+    public static var shared = URLObservable()
+}
+
+struct URLHandler {
+    public static var shared = URLHandler()
+
+    func processURL(url: URL) {
+        guard let urlComponenents = NSURLComponents(url: url, resolvingAgainstBaseURL: false),
+              let uriHost = urlComponenents.host,
+              let params = urlComponenents.queryItems else {
+                // Fall back to old url handler (for files)
+                if url.pathExtension == "ipa" {
+                    Installer.install(ipaUrl: url, export: false, returnCompletion: { _ in
+                    Task { @MainActor in
+                        AppsVM.shared.fetchApps()
+                        NotifyService.shared.notify(
+                            NSLocalizedString("notification.appInstalled", comment: ""),
+                            NSLocalizedString("notification.appInstalled.message", comment: "")
+                        )
+                    }})
+                }
+                return
+            }
+        // URI format: playcoverapp://<object>?action=<action>&<param>=<value>
+        // Example: playcoverapp://source?action=add&url=https://homebrew.playcover.io
+        // Switch case for main uri path
+        switch uriHost {
+        case "source":
+            processSourceURL(params: params)
+        default:
+            // Print URL to console and break
+            print("Unknown URL: \(url)")
+        }
+    }
+
+    func processSourceURL(params: [URLQueryItem]) {
+        if let actionParam = params[0].value {
+            URLObservable.shared.type = .source
+            switch actionParam {
+            case "add":
+                // Add source
+                if let url = params[1].value {
+                    URLObservable.shared.url = url
+                    URLObservable.shared.action = .add
+                }
+            case "remove":
+                // Remove source
+                if let url = params[1].value {
+                    URLObservable.shared.url = url
+                    URLObservable.shared.action = .remove
+                }
+            case "update":
```

Comment:
In the matter of `remove` and `update`, I don't see any use case that ends well. because anybody can pass a remove url and delete my source for example in users' playcover (or vice versa) or even i can pass decrypt.day url as my old url and tell playcover to replace it with mine.
i know that user must accept any changes but at actually is more harmful than useful.
i would say we don't need such vulnerabilities since all users are not expertise and they usually don't check what's going on when they trust their providers.
`add` is enough. that's all.

Reply:
That's reserved for future use. If you check, there is no code that makes use of `update` and `remove`. That is just prebuilt infrastructure in case we want to use it for other stuff in the code later.

Reply:
since we can never guarantee that it won't be abused, i would say to not put it there in the codes at all

Reply:
Again, it is added there as ready-built infrastructure. Other content (eg. keymaps or apps) can use it later on if needed (which they very well may be). It being there doesn't mean it will get implemented